### PR TITLE
util: Add utility function for binding CQ to EP

### PR DESCRIPTION
**Pull request by gladkovdmitry17**
The PR also includes:
Updates of rxm, udp and mlx providers to utilize new utility function

### CODE REVIEW

**Comment by gladkovdmitry17** [comment]
@alexander-sannikov @a-ilango @shefty Could you take a look into the PR, please? Thanks in advance.

**Comment by alexander-sannikov** [comment]
@gladkovdmitry17 looks good for me.

File: prov/util/src/util_ep.c [hunk 1]
```diff
@@ -36,6 +36,38 @@
...
+	}
+
+	if (flags & (FI_TRANSMIT | FI_RECV)) {
+		return fid_list_insert(&cq->ep_list,
+				       &cq->ep_list_lock,
+				       &ep->ep_fid.fid);
+	}
+
+	return FI_SUCCESS;
+}
```
Comment (shefty) [inline prov/util/src/util_ep.c]: Please move the debug print statements into this call.  I think you can access the provider structure from the ep, but if not, just pass in the provider structure to the call.

File: prov/udp/src/udpx_ep.c [hunk 2]
```diff
@@ -575,38 +575,30 @@ static int udpx_ep_close(struct fid *fid)
...
-	}
-
-	if (((flags & FI_TRANSMIT) && ep->util_ep.tx_cq) ||
-	    ((flags & FI_RECV) && ep->util_ep.rx_cq)) {
+	ret = ofi_ep_bind_cq(&ep->util_ep, cq, flags);
+	if (ret) {
 		FI_WARN(&udpx_prov, FI_LOG_EP_CTRL,
-			"duplicate CQ binding\n");
-		return -FI_EINVAL;
+			"CQ binding to EP is undable\n");
```
Comment (shefty) [inline prov/udp/src/udpx_ep.c]: See comment below.  We can drop the print here and move internal to bind, where we can report the reason for the failure.

File: prov/util/src/util_ep.c [hunk 3]
```diff
@@ -87,7 +119,6 @@ int ofi_endpoint_init(struct fid_domain *domain, const struct util_prov *util_pr
 	ep->caps = info->caps;
 	ep->domain = util_domain;
 	ep->progress = progress;
-	ep->caps = info->caps;
```
Comment (shefty) [inline prov/util/src/util_ep.c]: Why was this removed?

Comment (gladkovdmitry17) [inline prov/util/src/util_ep.c]: The statement is redundant (this is double set) the previous one is in 119 line of code

Comment (shefty) [inline prov/util/src/util_ep.c]: Oh, hey it is!  Thanks

File: prov/util/src/util_ep.c (hunk 1 above)
Comment (gladkovdmitry17) [inline prov/util/src/util_ep.c]: Thanks. Okay, will do that

File: prov/util/src/util_ep.c [hunk 4]
```diff
@@ -36,6 +36,38 @@
...
+		return -FI_EBADFLAGS;
+
+	if (((flags & FI_TRANSMIT) && ep->tx_cq) ||
+	    ((flags & FI_RECV) && ep->rx_cq))
+		return -FI_EINVAL;
+
+	if (flags & FI_TRANSMIT) {
+		ep->tx_cq = cq;
+		if (!(flags & FI_SELECTIVE_COMPLETION))
+			ep->tx_op_flags |= FI_COMPLETION;
```
Comment (a-ilango) [inline prov/util/src/util_ep.c]: These flags have to be initialized in ofi_endpoint_init from the info arg. Same for rx_op_flags below.

Comment (gladkovdmitry17) [inline prov/util/src/util_ep.c]: Sorry, my bad. Forgot to set it in initial value.
Thanks a lot, Arun! It is very good point.

**Comment by gladkovdmitry17** [comment]
@shefty @a-ilango I've updated the PR with proposed changes

**Comment by shefty** [comment]
> +	if (flags & FI_TRANSMIT) {
> +		ep->tx_cq = cq;
> +		if (!(flags & FI_SELECTIVE_COMPLETION))
> +			ep->tx_op_flags |= FI_COMPLETION;
> 
> These flags have to be initialized in ofi_endpoint_init from the info
> arg. Same for rx_op_flags below.

This setting is a little different.  It is indicating that all transmit operations must generate a completion.  Typically the app will not set tx_op_flags with FI_COMPLETION when they don't want selective completions.

I think we want to keep this, otherwise, what will the provider check to see if a completion should be generated or not.

**Comment by a-ilango** [comment]
@shefty Yes, we want to keep the current change as it is. What I meant was the op_flags needs to be initialized with op_flags from fi_info passed to ofi_endpoint_init.

**Comment by shefty** [comment]
Ah - got it.  Yes, I agree.

File: prov/util/src/util_ep.c [hunk 5]
```diff
@@ -36,6 +36,44 @@
...
+	}
+
+	if (flags & FI_RECV) {
+		ep->rx_cq = cq;
+		if (!(flags & FI_SELECTIVE_COMPLETION))
+			ep->rx_op_flags |= FI_COMPLETION;
+		ofi_atomic_inc32(&cq->ref);
+	}
+
+	if (flags & (FI_TRANSMIT | FI_RECV)) {
```
Comment (shefty) [inline prov/util/src/util_ep.c]: This if statement changes the behavior for the udp provider, which only adds the ep to the receive cq list.  That could be the reason why that provider is failing in travis.  (udp may only be cleaning up the receive cq.)  It results in slightly more overhead for udp, since the underlying socket will be processed for receives twice, once coming from the send cq.  For udp, please revert those changes and have it keep its own bind call.

To keep some code sharing, maybe separate out the bind validation check, so that this call and the udp call can both use them?

Comment (gladkovdmitry17) [inline prov/util/src/util_ep.c]: Ooop... Good point. Thanks. I've applied your remarks to the new version for the PR .

**Comment by gladkovdmitry17** [comment]
@shefty @a-ilango I've fixed up the problem with UDP as Sean proposed. Thanks for proposal.
Do you have any additional comments?

**Comment by a-ilango** [comment]
👍
